src/algebra/module/submodule_bilinear.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Eric Wieser
+-/
+import linear_algebra.span
+import linear_algebra.bilinear_map
+
+/-!
+# Images of pairs of submodules under bilinear maps
+
+This file provides `submodule.map₂`, which is later used to implement `submodule.has_mul`.
+
+## Main results
+
+* `submodule.map₂_eq_span_image2`: the image of two submodules under a bilinear map is the span of
+  their `set.image2`.
+
+## Notes
+
+This file is quite similar to the n-ary section of `data.set.basic` and to `order.filter.n_ary`.
+Please keep them in sync.
+-/
+
+universes uι u v
+
+open set
+open_locale big_operators
+open_locale pointwise
+
+namespace submodule
+
+variables {ι : Sort uι} {R M N P : Type*}
+variables [comm_semiring R] [add_comm_monoid M] [add_comm_monoid N] [add_comm_monoid P]
+variables [module R M] [module R N] [module R P]
+
+/-- Map a pair of submodules under a bilinear map.
+
+This is the submodule version of `set.image2`.  -/
+def map₂ (f : M →ₗ[R] N →ₗ[R] P) (p : submodule R M) (q : submodule R N) : submodule R P :=
+⨆ s : p, q.map $ f s
+
+theorem apply_mem_map₂ (f : M →ₗ[R] N →ₗ[R] P) {m : M} {n : N}
+  {p : submodule R M} {q : submodule R N} (hm : m ∈ p) (hn : n ∈ q) : f m n ∈ map₂ f p q :=
+(le_supr _ ⟨m, hm⟩ : _ ≤ map₂ f p q) ⟨n, hn, rfl⟩
+
+theorem map₂_le {f : M →ₗ[R] N →ₗ[R] P}
+  {p : submodule R M} {q : submodule R N} {r : submodule R P} :
+  map₂ f p q ≤ r ↔ ∀ (m ∈ p) (n ∈ q), f m n ∈ r :=
+⟨λ H m hm n hn, H $ apply_mem_map₂ _ hm hn,
+ λ H, supr_le $ λ ⟨m, hm⟩, map_le_iff_le_comap.2 $ λ n hn, H m hm n hn⟩
+
+variables R
+theorem map₂_span_span (f : M →ₗ[R] N →ₗ[R] P) (s : set M) (t : set N) :
+  map₂ f (span R s) (span R t) = span R (set.image2 (λ m n, f m n) s t) :=
+begin
+  apply le_antisymm,
+  { rw map₂_le, intros a ha b hb,
+    apply span_induction ha,
+    work_on_goal 1 { intros, apply span_induction hb,
+      work_on_goal 1 { intros, exact subset_span ⟨_, _, ‹_›, ‹_›, rfl⟩ } },
+    all_goals {
+      intros,
+      simp only [linear_map.map_zero, linear_map.zero_apply, zero_mem,
+        linear_map.map_add, linear_map.add_apply, linear_map.map_smul, linear_map.smul_apply] },
+    all_goals {
+      solve_by_elim [add_mem _ _, zero_mem _, smul_mem _ _ _]
+        { max_depth := 4, discharger := tactic.interactive.apply_instance } } },
+  { rw span_le, rintros _ ⟨a, b, ha, hb, rfl⟩,
+    exact apply_mem_map₂ _ (subset_span ha) (subset_span hb) }
+end
+variables {R}
```

Comment:
This refactor extracted from my previous branch introduces the more flexible lemmas `map₂_span_left/right` which may be useful elsewhere.
```suggestion
section
variables R (f : M →ₗ[R] N →ₗ[R] P) (s : set M) (t : set N)
theorem map₂_span_left :
  span R (set.image2 (λ m, f m) (span R s) t) = span R (s.image2 (λ m, f m) t) :=
span_eq_of_le _ (λ t ⟨m,n,hm,hn,h⟩, begin
  rw ← h, have := apply_mem_span_image_of_mem_span (f.flip n) hm,
  refine span_mono _ this, exact λ t ⟨m,hm,h⟩, ⟨m,n,hm,hn,h⟩,
end) $ span_mono $ set.image2_subset_right subset_span

theorem map₂_span_right :
  span R (s.image2 (λ m, f m) (span R t)) = span R (s.image2 (λ m, f m) t) :=
span_eq_of_le _ (λ t ⟨m,n,hm,hn,h⟩, begin
  rw ← h, refine span_mono _ (apply_mem_span_image_of_mem_span (f m) hn),
  exact λ t ⟨n,hn,h⟩, ⟨m,n,hm,hn,h⟩,
end) $ span_mono $ set.image2_subset_left subset_span

theorem map₂_span_span : map₂ f (span R s) (span R t) = span R (s.image2 (λ m, f m) t) :=
begin
  rw [← map₂_span_left, ← map₂_span_right], apply le_antisymm,
  { rw map₂_le, exact λ m hm n hn, subset_span ⟨m,n,hm,hn,rfl⟩ },
  { rw span_le, exact λ _ ⟨m,n,hm,hn,h⟩, h ▸ apply_mem_map₂ _ hm hn },
end
end
```

Reply:
I'd prefer to leave this PR as primarily moving things around, rather than refactoring any proofs. Those lemmas look good, but I'll let you add them in a follow-up.

Reply:
Do you prefer me do it in #13733 or another PR?

Reply:
Probably a separate PR makes sense, especially if you don't use the new lemmas anywhere in your existing PR.

Reply:
Looking more closely, I don't really like the look of these lemma statements - do we have anything analogous for the normal unary image?

Reply:
They're analogues of [submodule.span_span](https://leanprover-community.github.io/mathlib_docs/linear_algebra/span.html#submodule.span_span)

Reply:
Hmm, so probably the two lemmas should go to that file (`linear_algebra/span`)

Reply:
`span_span` doesn't have `image` in it, so it's not exactly an analogue. A closer analogue would be the following:
```
theorem span_image_span (f : M →ₗ[R] N) (s : set M) :
  span R (f '' (span R s)) = span R (f '' s) :=
by rw [← map_span, ← map_span, span_span]
```
The two `map_span` transform the goal to `map f (span R ↑(span R s)) = map f (span R s)`, which we can't do in the bilinear case, so there're some real difference here.

Reply:
Why is `R` made explicit here? Seems it can always be inferred from `f`?

Reply:
I've [created a branch](https://github.com/leanprover-community/mathlib/compare/tensor_finite'...tensor_finite?expand=1) with the two lemmas, ready to PR once this PR is merged; alternatively, I could merge it to #13733 if you approves.